### PR TITLE
[TOPIC-GPIO] Add LiteX GPIO driver

### PR DESCRIPTION
**Pull request by mateusz-holenko**
This adds LiteX GPIO driver and enables it in litex_vexriscv board.

### CODE REVIEW

File: CODEOWNERS
Comment (galak) [inline CODEOWNERS]: Can probably just make this `/drivers/gpio/*litex*`

File: dts/bindings/gpio/litex,gpio.yaml
Comment (galak) [inline dts/bindings/gpio/litex,gpio.yaml]: Maybe `nr-gpios` instead.

**Review by galak — DISMISSED** [review]
Minor change to DTS binding.

File: dts/bindings/gpio/litex,gpio.yaml
Comment (mateusz-holenko) [reply dts/bindings/gpio/litex,gpio.yaml]: It looks like properties with `gpios` in the name are handled in a special way by the DTS extractor script - see [here](https://github.com/zephyrproject-rtos/zephyr/blob/303ef27535cc7df590d76584bd0c0ee991d222fd/scripts/dts/extract_dts_includes.py#L87).

As a result after the suggested rename `sanitycheck` fails with: 
```
device tree error: bad phandle in <Property 'nr-gpios' at '/soc/gpio@e0003000' in 'litex_vexriscv.dts.pre.tmp'>
```

Comment (galak) [reply dts/bindings/gpio/litex,gpio.yaml]: I think this is because the type of property is not correct.  It should not be 'array' but 'int'

File: dts/bindings/gpio/litex,gpio.yaml
Comment (galak) [inline dts/bindings/gpio/litex,gpio.yaml]: This should be `type: int`

File: dts/bindings/gpio/litex,gpio.yaml
Comment (mateusz-holenko) [reply dts/bindings/gpio/litex,gpio.yaml]: Changing the type to `int` did not help.

I noticed that writing `nr-gpios: <&intc0>` in DTS solves the problem of "bad phandle", but (apart from not having any sense in this context) it creates another one:

```
device tree error: missing data after phandle in <Property 'nr-gpios' at '/soc/gpio@e0003000' in 'litex_vexriscv.dts.pre.tmp'>
```

Reverting type to `array` and writing `nr-gpios <&intc0 1 1>` solves parsing problem, but then `gen_defines.py` script fails in `write_gpio` function.

I have a strong feeling that properties with "gpios" in the name are expected to reference interrupt controller and are used in the interrupts context.

This is not what we intend for LiteX GPIO port.

Comment (galak) [reply dts/bindings/gpio/litex,gpio.yaml]: Can you push the PR with the change to it being 'int' and that shows the CI error.  It will be easier for me to take a look to see what's going on than.

File: dts/bindings/gpio/litex,gpio.yaml
Comment (mateusz-holenko) [reply dts/bindings/gpio/litex,gpio.yaml]: Fixed.

File: CODEOWNERS
Comment (mateusz-holenko) [reply CODEOWNERS]: Changed.

File: dts/bindings/gpio/litex,gpio.yaml
Comment (mateusz-holenko) [reply dts/bindings/gpio/litex,gpio.yaml]: @galak Pushed.

**Comment by galak** [comment]
You'll need the fix in:

https://github.com/zephyrproject-rtos/zephyr/pull/18089

Also, instead of `nrgpios` we need to make `ngpios` (sorry for the churn on that).

**Comment by galak** [comment]
> You'll need the fix in:
> 
> #18089
> 
> Also, instead of `nrgpios` we need to make `ngpios` (sorry for the churn on that).

Re-based and fixed up the yaml to be `ngpios`

**Comment by mateusz-holenko** [comment]
> > You'll need the fix in:
> > #18089
> > Also, instead of `nrgpios` we need to make `ngpios` (sorry for the churn on that).
> 
> Re-based and fixed up the yaml to be `ngpios`

Thanks.

**Comment by mateusz-holenko** [comment]
I fixed the base addresses of gpio ports in the DTS.

**Comment by mateusz-holenko** [comment]
@galak Is there anything else I should do to have this PR merged?

**Comment by mateusz-holenko** [comment]
@galak The code is rebased on top of the 2.0.0 release and all requested changes are applied. Can we merge it?

**Comment by mateusz-holenko** [comment]
Rebased to resolve conflicts after merging #18649.

**Comment by mateusz-holenko** [comment]
@galak Is there something blocking this PR from being merged?

**Comment by galak** [comment]
> @galak Is there something blocking this PR from being merged?

We should migrate it to the new GPIO apis in the 'topic-gpio' branch.  See issue #18530

**Comment by mateusz-holenko** [comment]
@galak The driver is now rewritten to use the new GPIO api.

**Comment by mnkp** [comment]
It looks like the driver is still using the old GPIO API. This PR needs to target `topic-gpio` branch.

**Comment by mateusz-holenko** [comment]
@mnkp You are right - I had pushed a wrong branch. Sorry for the confussion.

It's fixed now - the current version is using the new GPIO API and the PR is now targetting `topic-gpio` branch.

**Review by mnkp — CHANGES_REQUESTED** [review]
It seems like this PR is missing updates of the soc .dts file.

File: drivers/gpio/gpio_litex.c
Comment (PiotrZierhoffer) [reply drivers/gpio/gpio_litex.c]: @mnkp I can't find any information if there will be other options for `access_op`. Should the driver still filter for `GPIO_ACCESS_BY_PIN`?

File: drivers/gpio/gpio_litex.c
Comment (PiotrZierhoffer) [reply drivers/gpio/gpio_litex.c]: Should `manage_callback` rely on `gpio_manage_callback` and just never call `gpio_fire_callback`? This controller doesn't support interrupts at all, so maybe `-ENOTSUP` would make sense here as well?

**Comment by PiotrZierhoffer** [comment]
@mnkp you mention lack of changes to the soc dts. I believe the relevant changes are in riscv32-litex-vexriscv.dtsi. Am I missing something?

**Comment by mnkp** [comment]
> @mnkp you mention lack of changes to the soc dts. I believe the relevant changes are in riscv32-litex-vexriscv.dtsi. Am I missing something?

Yes, they are. Sorry, I somehow overlooked the file.

File: dts/bindings/gpio/litex,gpio.yaml
Comment (PiotrZierhoffer) [reply dts/bindings/gpio/litex,gpio.yaml]: Changed to bool

File: drivers/gpio/gpio_litex.c
Comment (PiotrZierhoffer) [reply drivers/gpio/gpio_litex.c]: Fixed

File: drivers/gpio/gpio_litex.c
Comment (PiotrZierhoffer) [reply drivers/gpio/gpio_litex.c]: Fixed, following the pattern used in other controllers

File: drivers/gpio/gpio_litex.c
Comment (PiotrZierhoffer) [reply drivers/gpio/gpio_litex.c]: fixed

File: drivers/gpio/gpio_litex.c
Comment (PiotrZierhoffer) [reply drivers/gpio/gpio_litex.c]: removed

File: drivers/gpio/gpio_litex.c
Comment (PiotrZierhoffer) [reply drivers/gpio/gpio_litex.c]: `pin_interrupt_configure` implemented with `-ENOTSUP`

File: drivers/gpio/gpio_litex.c
Comment (PiotrZierhoffer) [reply drivers/gpio/gpio_litex.c]: Done.

**Comment by PiotrZierhoffer** [comment]
@mnkp the requested changes are implemented, tested and pushed.

**Review by mnkp — APPROVED** [review]

**Comment by galak** [comment]
Can you rebase this on `topic-gpio` thanks.

**Comment by mateusz-holenko** [comment]
@galak It's rebased

**Comment by carlescufi** [comment]
@galak can you take another look?

**Comment by carlescufi** [comment]
@galak @pabigot can we merge this?

File: drivers/gpio/gpio_litex.c
Comment (pabigot) [inline drivers/gpio/gpio_litex.c]: The path to this line through `gpio_pin_write()` appears to shift an unvalidated 32-bit value into the bits at and above the provided bit.  Since this is deprecated API I suppose that might not matter.  A simple solution would be to change the parameter type from `u32_t` to `bool`.

**Review by pabigot — COMMENTED** [review]
I don't have this board and haven't inspected this closely, and a quick search didn't locate any documentation (was curious about only having 4 GPIO signals supported).

Assuming it passes both tests the only thing that I'd recommend doing is below.  I won't insist on it since the issue's likely to be fixed before the topic branch merges.

File: drivers/gpio/gpio_litex.c
Comment (mateusz-holenko) [reply drivers/gpio/gpio_litex.c]: Changed `val` argument type to `bool`.

Comment (mateusz-holenko) [reply drivers/gpio/gpio_litex.c]: @pabigot Is it good now?

Comment (pabigot) [reply drivers/gpio/gpio_litex.c]: Yes, thanks.

File: dts/bindings/gpio/litex,gpio.yaml
Comment (galak) [inline dts/bindings/gpio/litex,gpio.yaml]: remove `type` and `description`, should be inherited from gpio-controller.yaml

**Review by galak — CHANGES_REQUESTED** [review]
minor dts binding update needed.

File: dts/bindings/gpio/litex,gpio.yaml
Comment (mateusz-holenko) [reply dts/bindings/gpio/litex,gpio.yaml]: removed

**Review by galak — APPROVED** [review]

**Pull request merged**
